File: backend/app/schemas/ad_discovery.py
"""Pydantic schemas for Ad Discovery feature."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator


def _validate_domain(v: str) -> str:
    """Reject domains containing '..' sequences."""
    if ".." in v:
        raise ValueError("Domain must not contain '..' sequences")
    return v


# ── Patterns ────────────────────────────────────────────────────────

class PatternEntry(BaseModel):
    id: int
    pattern: str
    is_regex: bool
    weight: float
    category: str
    is_default: bool
    enabled: bool

    model_config = {"from_attributes": True}


class PatternCreateRequest(BaseModel):
    pattern: str = Field(..., min_length=1, max_length=253)
    is_regex: bool = False
    weight: float = Field(0.5, ge=0.1, le=1.0)
    category: str = Field(..., min_length=1, max_length=50,
                          pattern="^(ads|tracking|telemetry|analytics|fingerprinting|custom)$")


class PatternUpdateRequest(BaseModel):
    weight: Optional[float] = Field(None, ge=0.1, le=1.0)
    enabled: Optional[bool] = None
    category: Optional[str] = Field(None, pattern="^(ads|tracking|telemetry|analytics|fingerprinting|custom)$")


class PatternListResponse(BaseModel):
    patterns: list[PatternEntry] = []


# ── Reference Lists ─────────────────────────────────────────────────

class ReferenceListEntry(BaseModel):
    id: int
    name: str
    url: str
    is_default: bool
    enabled: bool
    domain_count: int
    last_fetched_at: Optional[datetime] = None
    fetch_interval_hours: int
    last_error: Optional[str] = None
    last_error_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class ReferenceListCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    url: str = Field(..., min_length=10, max_length=2000, pattern="^https://")
    fetch_interval_hours: int = Field(24, ge=1, le=168)


class ReferenceListUpdateRequest(BaseModel):
    enabled: Optional[bool] = None
    fetch_interval_hours: Optional[int] = Field(None, ge=1, le=168)


class ReferenceListResponse(BaseModel):
    lists: list[ReferenceListEntry] = []


# ── Suspects ────────────────────────────────────────────────────────

class SuspectEntry(BaseModel):
    id: int
    domain: str
    first_seen_at: datetime
    last_seen_at: datetime
    query_count: int
    heuristic_score: float
    matched_patterns: Optional[list] = None
    community_hits: int
    community_lists: Optional[list] = None
    source: str
    status: str
    resolved_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class SuspectListResponse(BaseModel):
    suspects: list[SuspectEntry] = []
    total: int = 0
    page: int = 1
    page_size: int = 50


class SuspectStatusUpdate(BaseModel):
    status: str = Field(..., pattern="^(confirmed|dismissed|blocked)$")


class SuspectManualAdd(BaseModel):
    domain: str = Field(..., min_length=1, max_length=253)
    _validate_domain = field_validator("domain")(_validate_domain)


class SuspectBlockRequest(BaseModel):
    domain: str = Field(..., min_length=1, max_length=253)
    _validate_domain = field_validator("domain")(_validate_domain)
    target: str = Field(..., pattern="^(deny_list|custom_list)$")
    list_id: Optional[int] = None


class SuspectBulkActionRequest(BaseModel):
    domains: list[str] = Field(..., min_length=1, max_length=100)

    @field_validator("domains")
    @classmethod
    def validate_domains(cls, v: list[str]) -> list[str]:
        for d in v:
            _validate_domain(d)
        return v
    action: str = Field(..., pattern="^(block|dismiss|confirm)$")
    target: Optional[str] = Field(None, pattern="^(deny_list|custom_list)$")
    list_id: Optional[int] = None


# ── Custom Lists ────────────────────────────────────────────────────

class CustomListEntry(BaseModel):
    id: int
    name: str
    description: str
    domain_count: int
    created_at: datetime
    updated_at: datetime
    deployed: bool
    adlist_url: Optional[str] = None

    model_config = {"from_attributes": True}


class CustomListCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field("", max_length=500)


class CustomListUpdateRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=500)


class CustomListResponse(BaseModel):
    lists: list[CustomListEntry] = []


class CustomListDomainEntry(BaseModel):
    id: int
    domain: str
    added_at: datetime
    comment: str = ""

    model_config = {"from_attributes": True}


class CustomListDomainsResponse(BaseModel):
    domains: list[CustomListDomainEntry] = []
    total: int = 0
    page: int = 1
    page_size: int = 100


class CustomListAddDomainsRequest(BaseModel):
    domains: list[str] = Field(..., min_length=1, max_length=100)
    comment: str = Field("", max_length=500)

    @field_validator("domains")
    @classmethod
    def validate_domains(cls, v: list[str]) -> list[str]:
        for d in v:
            _validate_domain(d)
        return v


# ── Analysis ────────────────────────────────────────────────────────

class AnalyzeRequest(BaseModel):
    period: str = Field("24h", pattern="^(24h|7d|30d)$")
    min_score: float = Field(0.15, ge=0.0, le=1.0)


class AnalyzeResponse(BaseModel):
    new_suspects: int = 0
    updated_suspects: int = 0
    total_domains_analyzed: int = 0


# ── Status & Config ─────────────────────────────────────────────────

class AdDiscoveryStatusResponse(BaseModel):
    suspects_new: int = 0
    suspects_confirmed: int = 0
    suspects_dismissed: int = 0
    suspects_blocked: int = 0
    last_analysis_at: Optional[datetime] = None
    background_task_running: bool = False
    reference_lists_active: int = 0
    reference_lists_total: int = 0
    custom_lists_total: int = 0
    custom_lists_deployed: int = 0


class AdDiscoveryConfigResponse(BaseModel):
    background_interval_hours: int = 6
    heuristic_weight: float = 0.4
    community_weight: float = 0.6
    min_score: float = 0.15
    re_evaluation_threshold: float = 0.3
    background_enabled: bool = True


class AdDiscoveryConfigUpdate(BaseModel):
    background_interval_hours: Optional[int] = Field(None, ge=1, le=24)
    heuristic_weight: Optional[float] = Field(None, ge=0.0, le=1.0)
    community_weight: Optional[float] = Field(None, ge=0.0, le=1.0)
    min_score: Optional[float] = Field(None, ge=0.0, le=1.0)
    re_evaluation_threshold: Optional[float] = Field(None, ge=0.0, le=1.0)
    background_enabled: Optional[bool] = None

File: client/src/api/adDiscovery.ts
/**
 * API client for Ad Discovery feature.
 */
import { apiClient } from '../lib/api';

// ── Types ────────────────────────────────────────────────────────

export interface SuspectEntry {
  id: number;
  domain: string;
  first_seen_at: string;
  last_seen_at: string;
  query_count: number;
  heuristic_score: number;
  matched_patterns: string[] | null;
  community_hits: number;
  community_lists: string[] | null;
  source: string;
  status: string;
  resolved_at: string | null;
}

export interface PatternEntry {
  id: number;
  pattern: string;
  is_regex: boolean;
  weight: number;
  category: string;
  is_default: boolean;
  enabled: boolean;
}

export interface ReferenceListEntry {
  id: number;
  name: string;
  url: string;
  is_default: boolean;
  enabled: boolean;
  domain_count: number;
  last_fetched_at: string | null;
  fetch_interval_hours: number;
  last_error: string | null;
}

export interface CustomListEntry {
  id: number;
  name: string;
  description: string;
  domain_count: number;
  created_at: string;
  updated_at: string;
  deployed: boolean;
  adlist_url: string | null;
}

export interface CustomListDomainEntry {
  id: number;
  domain: string;
  added_at: string;
  comment: string;
}

export interface AdDiscoveryStatus {
  suspects_new: number;
  suspects_confirmed: number;
  suspects_dismissed: number;
  suspects_blocked: number;
  last_analysis_at: string | null;
  background_task_running: boolean;
  reference_lists_active: number;
  reference_lists_total: number;
  custom_lists_total: number;
  custom_lists_deployed: number;
}

export interface AdDiscoveryConfig {
  background_interval_hours: number;
  heuristic_weight: number;
  community_weight: number;
  min_score: number;
  re_evaluation_threshold: number;
  background_enabled: boolean;
}

// ── API Functions ────────────────────────────────────────────────

const BASE = '/api/pihole/ad-discovery';

// Status & Config
export async function getAdDiscoveryStatus(): Promise<AdDiscoveryStatus> {
  const { data } = await apiClient.get(`${BASE}/status`);
  return data;
}

export async function getAdDiscoveryConfig(): Promise<AdDiscoveryConfig> {
  const { data } = await apiClient.get(`${BASE}/config`);
  return data;
}

export async function updateAdDiscoveryConfig(update: Partial<AdDiscoveryConfig>): Promise<AdDiscoveryConfig> {
  const { data } = await apiClient.patch(`${BASE}/config`, update);
  return data;
}

// Analysis
export async function startAnalysis(period = '24h', minScore = 0.15) {
  const { data } = await apiClient.post(`${BASE}/analyze`, { period, min_score: minScore });
  return data;
}

// Suspects
export async function getSuspects(params: {
  status?: string; source?: string; sort_by?: string;
  page?: number; page_size?: number;
}) {
  const { data } = await apiClient.get(`${BASE}/suspects`, { params });
  return data as { suspects: SuspectEntry[]; total: number; page: number; page_size: number };
}

export async function updateSuspectStatus(domain: string, status: string) {
  const { data } = await apiClient.patch(`${BASE}/suspects/${encodeURIComponent(domain)}`, { status });
  return data;
}

export async function addManualSuspect(domain: string) {
  const { data } = await apiClient.post(`${BASE}/suspects/manual`, { domain });
  return data;
}

export async function blockSuspect(domain: string, target: string, listId?: number) {
  const { data } = await apiClient.post(`${BASE}/suspects/block`, {
    domain, target, list_id: listId,
  });
  return data;
}

export async function bulkAction(domains: string[], action: string, target?: string, listId?: number) {
  const { data } = await apiClient.post(`${BASE}/suspects/bulk-action`, {
    domains, action, target, list_id: listId,
  });
  return data;
}

// Patterns
export async function getPatterns() {
  const { data } = await apiClient.get(`${BASE}/patterns`);
  return data as { patterns: PatternEntry[] };
}

export async function createPattern(pattern: { pattern: string; is_regex: boolean; weight: number; category: string }) {
  const { data } = await apiClient.post(`${BASE}/patterns`, pattern);
  return data;
}

export async function updatePattern(id: number, update: { weight?: number; enabled?: boolean; category?: string }) {
  const { data } = await apiClient.patch(`${BASE}/patterns/${id}`, update);
  return data;
}

export async function deletePattern(id: number) {
  const { data } = await apiClient.delete(`${BASE}/patterns/${id}`);
  return data;
}

// Reference Lists
export async function getReferenceLists() {
  const { data } = await apiClient.get(`${BASE}/reference-lists`);
  return data as { lists: ReferenceListEntry[] };
}

export async function createReferenceList(list: { name: string; url: string; fetch_interval_hours?: number }) {
  const { data } = await apiClient.post(`${BASE}/reference-lists`, list);
  return data;
}

export async function updateReferenceList(id: number, update: { enabled?: boolean; fetch_interval_hours?: number }) {
  const { data } = await apiClient.patch(`${BASE}/reference-lists/${id}`, update);
  return data;
}

export async function deleteReferenceList(id: number) {
  const { data } = await apiClient.delete(`${BASE}/reference-lists/${id}`);
  return data;
}

export async function refreshReferenceLists() {
  const { data } = await apiClient.post(`${BASE}/reference-lists/refresh`);
  return data;
}

// Custom Lists
export async function getCustomLists() {
  const { data } = await apiClient.get(`${BASE}/custom-lists`);
  return data as { lists: CustomListEntry[] };
}

export async function createCustomList(list: { name: string; description?: string }) {
  const { data } = await apiClient.post(`${BASE}/custom-lists`, list);
  return data;
}

export async function updateCustomList(id: number, update: { name?: string; description?: string }) {
  const { data } = await apiClient.patch(`${BASE}/custom-lists/${id}`, update);
  return data;
}

export async function deleteCustomList(id: number) {
  const { data } = await apiClient.delete(`${BASE}/custom-lists/${id}`);
  return data;
}

export async function getCustomListDomains(id: number, page = 1, pageSize = 100) {
  const { data } = await apiClient.get(`${BASE}/custom-lists/${id}/domains`, {
    params: { page, page_size: pageSize },
  });
  return data as { domains: CustomListDomainEntry[]; total: number };
}

export async function addCustomListDomains(id: number, domains: string[], comment = '') {
  const { data } = await apiClient.post(`${BASE}/custom-lists/${id}/domains`, { domains, comment });
  return data;
}

export async function removeCustomListDomain(id: number, domain: string) {
  const { data } = await apiClient.delete(`${BASE}/custom-lists/${id}/domains/${encodeURIComponent(domain)}`);
  return data;
}

export async function deployCustomList(id: number) {
  const { data } = await apiClient.post(`${BASE}/custom-lists/${id}/deploy`);
  return data;
}

export async function undeployCustomList(id: number) {
  const { data } = await apiClient.post(`${BASE}/custom-lists/${id}/undeploy`);
  return data;
}

export async function exportCustomList(id: number): Promise<Blob> {
  const { data } = await apiClient.get(`${BASE}/custom-lists/${id}/export`, {
    responseType: 'blob',
  });
  return data;
}
